Demos update CmakeLists.txt

Wovchena opened the pull request:
There is no special reason for this PR

alalek commented on demos/CMakeLists.txt:
```diff
@@ -4,17 +4,20 @@
+if(NOT GENERATOR_IS_MULTI_CONFIG AND NOT DEFINED CMAKE_BUILD_TYPE)
+    message(STATUS "CMAKE_BUILD_TYPE not defined, 'Release' will be used")
+    # Setting CMAKE_BUILD_TYPE as CACHE must go before project(). Otherwise project() sets its value and set() doesn't take an effect
+    set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build, options are: None Debug Release RelWithDebInfo MinSizeRel ...")
```
> RelWithDebInfo MinSizeRel

Do all dependencies support these modes properly? Does it checked somewhere?

Wovchena replied:
I don't know. This message is what `CMAKE_BUILD_TYPE` usually has, I just put a copy of it to mimic it. Would it be better if I truncated the message to `Choose the type of build`?

IRDonch replied:
What was the purpose of this change, anyway? Wouldn't it be sufficient to just add `NOT GENERATOR_IS_MULTI_CONFIG` to the original `if` command?

Wovchena replied:
Setting `CMAKE_BUILD_TYPE` as CACHE puts the correct value into cache. Earlier version shadows original `CMAKE_BUILD_TYPE` cache var.

IRDonch replied:
> Earlier version shadows original CMAKE_BUILD_TYPE cache var.

Yeah, but... is it actually a problem?

Wovchena replied:
This resulted in having two wars. While cached value claimed the type is `""`, as well as cache viewers did, in fact Release was generated.

IRDonch replied:
Yeah, but again... is that actually a problem? We're effectively just "symlinking" the blank configuration to the release configuration.

Wovchena replied:
Blank configuration seems to be a separate valid value, decoupling these two should be cleaner.

IRDonch replied:
Yeah, in the blank configuration none of the `*_DEBUG` or `*_RELEASE` variables are used, resulting in a build with neither optimizations nor debug info. :-)

I'm not convinced that we need to support such a configuration, but I also don't care enough to argue about it, so keep it like that if you want.

alalek commented on demos/CMakeLists.txt:
```diff
@@ -103,33 +90,17 @@ endfunction()
+#     [HEADERS <header files>]
+#     [INCLUDE_DIRECTORIES <include dir>]
+#     [DEPENDENCIES <dependencies>])
+macro(add_demo)
```
omz_add_demo?

Wovchena replied:
Why? I don't see a reason to stay aligned with style of the old name. Plus this will conflict with `IE_SAMPLE*` naming which I didn't touch here.

vladimir-dudnik commented:
@Wovchena this seems to be many changes across demo CMakeLists files, could you please add to PR description what it will serve for? Basically why these changes are needed?

Wovchena commented:
> this seems to be many changes across demo CMakeLists files

I've split PR on multiple commits because of that.

> could you please add to PR description what it will serve for? Basically why these changes are needed?

There is no special reason for this PR, I just updated cmake usage. PR's name reflects exactly that. Regarding ideas staying behind each particular change, see commit messages.

vladimir-dudnik commented:
@Wovchena PR description is place to put summary info to free up people from looking at each particular commit to deduce what is the chages and whythey are needed. In this case, PR name says what you change (CMakeLists for demos) but it does not say why changes are needed, what is intention.

Wovchena commented:
I believe `update CmakeLists.txt` is the best summary. As I said, there is no special reason for this PR. What do you want me to put there?

Wovchena commented:
Jenkins please retry a build

Wovchena commented:
Jenkins please retry a build

IRDonch commented on demos/CMakeLists.txt:
```diff
@@ -103,33 +89,17 @@ endfunction()
+#     [DEPENDENCIES <dependencies>])
+macro(add_demo)
     set(oneValueArgs NAME)
     set(multiValueArgs SOURCES HEADERS DEPENDENCIES OPENCV_DEPENDENCIES INCLUDE_DIRECTORIES)
```
Shouldn't `OPENCV_DEPENDENCIES` be removed?

IRDonch commented:
I'm gonna rerun the builds, in case something was merged in the intervening days that broke them.

IRDonch commented:
Jenkins please retry a build
